src/irr/asset/COBJMeshFileLoader.cpp

```diff
@@ -39,13 +39,361 @@ namespace std
 }
 */
 
-#define NEW_SHADERS
+namespace
+{
+    constexpr const char* VERT_SHADER_NO_UV = 
+R"(#version 430 core
+
+layout (location = 0) in vec3 vPos;
+layout (location = 3) in vec3 vNormal;
+
+layout (location = 0) out vec3 LocalPos;
+layout (location = 1) out vec3 ViewPos;
+layout (location = 2) out vec3 Normal;
+
+layout (set = 1, binding = 0, row_major, std140) uniform UBO {
+    mat4 MVP;
+    mat4 MV;
+    mat3 NormalMat;
+    vec3 EyePos;
+} CamData;
+
+void main()
+{
+    LocalPos = vPos;
+    gl_Position = CamData.MVP*vec4(vPos, 1.0);
+    ViewPos = (CamData.MV*vec4(vPos, 1.0)).xyz;
+    Normal = normalize(CamData.NormalMat*vNormal);
+}
+)";
+    constexpr const char* FRAG_SHADER_NO_UV =
+R"(#version 430 core
+
+layout (location = 0) in vec3 LocalPos;
+layout (location = 1) in vec3 ViewPos;
+layout (location = 2) in vec3 Normal;
+layout (location = 0) out vec4 OutColor;
+
+#define ILLUM_MODEL_MASK 0x0fu
+layout (push_constant) uniform Block {
+    vec3 Ka;
+    vec3 Kd;
+    vec3 Ks;
+    vec3 Ke;
+    vec4 Tf;//w component doesnt matter
+    float Ns;
+    float d;
+    float bm;
+    float Ni;
+    float roughness;
+    float metallic;
+    float sheen;
+    float clearcoatThickness;
+    float clearcoatRoughness;
+    float anisotropy;
+    float anisoRotation;
+    //extra info
+    uint extra;
+} PC;
+
+#include <irr/builtin/glsl/brdf/specular/fresnel/fresnel.glsl>
+
+#define Ia 0.1
+void main()
+{
+    vec3 N = normalize(Normal);
+    vec3 L = normalize(-ViewPos);
+    vec3 R = -reflect(L,N);
+    float NdotL = max(dot(N,L), 0.0);
+    float VdotR = max(dot(L,R), 0.0);
```

Comment:
there should be a built-in for that and a header with utility functions, see
https://github.com/buildaworldnet/IrrlichtBAW/issues/384#issuecomment-546903906

Reply:
i dont get it. Builtin for normalizing and dot products?

Reply:
yes, cause there are tricks to get them faster, and user (all implementations and shaders) should rely on the function to get all the dot products and vectors (we make an improvement, every shader benefits, instead of non-DRY code).

Reply:
Yes, because there's hacks that speed it all up, and you dont want to keep code DRY